### DISCUSSION TITLE
python Communicator in parameter variation

### DISCUSSION THREAD

**Discussion by BLindeboom27** [the-anylogic-company/AnyLogic-Pypeline#24] [Q&A] [answered]
Hi,

Thanks again for your help previously. I am trying to create a parameter variation based on a single python script run. I manually added a variable called 'pyCommunicator ' of the type other 'PyCommunicator' with initial value new 'PyCommunicator()'.   help

In the 'initial experiment set' block of the properties section, I import the python file via 'pyCommunicator.run("import GA");' 
However, when I try to run the parameter variation I get the error code "PyCommunicator.java:669' indicating that the source code is not present.  It also states '669 is not a valid line number in com.anylogic.libraries.pypeline.PyCommunicator'

When I click on the error I am redirected to a class file editor. Here I add the jar file at the 'Change Attached Source' option, however the error does not change. The jar file is in the same folder as the anylogic and python files.  Do you by any chance now if I missing something here? 

The python communicator does work for a normal simulation (using the Pipeline block).

**Comment by t-wolfeadam** [comment, marked as answer]
Hello! To use the PyCommunicator from the experiment screen - in addition to adding the variable like you did - you'll need to instantiate and start it before each simulation run (in the relevant callback field). This is because the PyCommunicator is actually an agent under the hood and can't persist on its own between runs.

You can see an example of this in the following demo model: [Demo - Python from experiment screen.zip](https://github.com/t-wolfeadam/AnyLogic-Pypeline/files/7809661/Demo.-.Python.from.experiment.screen.zip)

If you need to maintain information across runs (e.g., trying to run a python based optimization), you'll either need to record data to some file at a run's end and read it at the start of the next, or use a more appropriate method of using Python to control a model (e.g., Cloud API or Alpyne).

One last thing for your reference: the message you see in the console shows the sequence of function calls leading up to the error happening; `PyCommunicator.java:669` is not an error code but rather the location of one of those function calls ("line 669 of the PyCommunicator.java file"). In the future, pasting the full error log would help the most with troubleshooting.

**Reply by BLindeboom27** [reply]
Hi there, thanks for the quick reply and the explanation.  I indeed have to maintain information across but thought it possible to store the values of a singular run and use them for other runs also. I will adopt your solution of storing the results first in a separate file, thank you.

**Reply by t-wolfeadam** [reply]
One warning about that approach: if you have _"Allow parallel evaluations"_ enabled, you may encounter issues with parallel file reading/writing or sometimes have a `Socket Closed` runtime error for the last simulation run (it happens if you try to call Python code from the "After simulation run" callback). I would suggest verifying your results in this case, or just use sequential runs altogether 🙂
